File: Java DSA/MergeSort.java
import java.util.*;

public class MergeSort {
    int[] merge_sort(int[] arr) {
        int l = arr.length;
        if (l <= 1) return arr;
        int mid = l / 2;
        int[] left = new int[mid];
        int[] right = new int[l - mid];
        for (int i = 0; i < mid; i++)
            left[i] = arr[i];
        left = merge_sort(left);
        for(int i = mid; i < l; i++)
            right[i-mid] = arr[i];
        right = merge_sort(right);
        return merge(left, right);
    }
    int[] merge(int[] left, int[] right) {
        int[] result = new int[left.length + right.length];
        int i = 0, j = 0,indx=0;
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j])
                result[indx++] = left[i++];
            else
                result[indx++] = right[j++];
        }
        while( i < left.length)
            result[indx++] = left[i++];
        while(j < right.length)
            result[indx++] = right[j++];
        return result;
    }
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size of the array");
        int length = sc.nextInt();
        int[] arr = new int[length];
        System.out.println("Enter the elements of the array");
        for (int i = 0; i < length; i++)
            arr[i] = sc.nextInt();
        MergeSort ob = new MergeSort();
        int[] result = ob.merge_sort(arr);
        System.out.println("Sorted array");
        for(int i = 0; i < result.length; i++)
            System.out.print(result[i] + " ");
    }
}

File: C++_DSA/MergeSort.cpp
#include <bits/stdc++.h>
using namespace std;

void merge(vector<int> &arr, int low, int mid, int high){
    int n1 = mid - low + 1;
    int n2 = high - mid;
    vector<int> left(n1), right(n2);
    for(int i=0;i<n1;i++)
        left[i] = arr[low+i];
    for(int i=0;i<n2;i++)
        right[i] = arr[mid+1+i];
    int i=0,j=0,k=low;
    while(i<n1 && j<n2){
        if(left[i]<=right[j]){
            arr[k] = left[i];
            i++;
        }
        else{
            arr[k] = right[j];
            j++;
        }
        k++;
    }
    while(i<n1){
        arr[k] = left[i];
        i++;
        k++;
    }
    while(j<n2){
        arr[k] = right[j];
        j++;
        k++;
    }
}

void mergeSort(vector<int> &arr, int low, int high){
    if(low<high){
        int mid = low + (high-low)/2;
        mergeSort(arr, low, mid);
        mergeSort(arr, mid+1, high);
        merge(arr, low, mid, high);
    }
}

int main(){
    vector<int> nums = {10, 7, 8, 9, 1, 5};
    mergeSort(nums, 0, nums.size()-1);
    for(int i=0;i<nums.size();i++){
        cout<<nums[i]<<" ";
    }
}